networkit/cpp/graph/GraphBuilder.cpp

```diff
@@ -12,86 +12,163 @@
 
 namespace NetworKit {
 
-GraphBuilder::GraphBuilder(count n, bool weighted, bool directed)
-    : n(n), selfloops(0), weighted(weighted), directed(directed), outEdges(n),
-      outEdgeWeights(weighted ? n : 0), inEdges(directed ? n : 0),
-      inEdgeWeights((directed && weighted) ? n : 0) {}
+GraphBuilder::GraphBuilder(count n, bool weighted, bool directed, bool useWholeEdges)
+    : n(n), selfloops(0), weighted(weighted), directed(directed), useWholeEdges(useWholeEdges) {
+    // pre-allocating the adjacency vectors
+    int max_threads = omp_get_max_threads();
+    outEdgesPerThread.resize(max_threads);
+    inEdgesPerThread.resize(max_threads);
+    for (int i = 0; i < max_threads; i++) {
+        outEdgesPerThread[i].resize(max_threads);
+        inEdgesPerThread[i].resize(max_threads);
+    }
+    if (weighted) {
+        outEdgeWeightsPerThread.resize(max_threads);
+        inEdgeWeightsPerThread.resize(max_threads);
+        for (int i = 0; i < max_threads; i++) {
+            outEdgeWeightsPerThread[i].resize(max_threads);
+            inEdgeWeightsPerThread[i].resize(max_threads);
+        }
+    }
+}
 
 void GraphBuilder::reset(count n) {
     this->n = n;
     selfloops = 0;
-    outEdges.assign(n, std::vector<node>{});
-    outEdgeWeights.assign(isWeighted() ? n : 0, std::vector<edgeweight>{}),
-        inEdges.assign(isDirected() ? n : 0, std::vector<node>{}),
-        inEdgeWeights.assign((isDirected() && isWeighted()) ? n : 0, std::vector<edgeweight>{});
+
+    int max_threads = omp_get_max_threads();
+    outEdgesPerThread.assign(max_threads, std::vector<std::vector<HalfEdge>>{});
+    outEdgeWeightsPerThread.assign(isWeighted() ? max_threads : 0,
+                                   std::vector<std::vector<edgeweight>>{}),
```

Comment:
Why not
```suggestion
    if (isWeighted()) {
        outEdgeWeightsPerThread.resize(max_threads);
    }
```

Also for the other two below.

Reply:
Only resizing does not clear the data structure (as the size is the same as before), therefore, we clear the data structure by assigning new empty vectors to ensure proper resetting.